vms/platformvm/network/warp.go

```diff
@@ -0,0 +1,356 @@
+func (s signatureRequestVerifier) verifySubnetValidatorCanNotValidate(
+	validationID ids.ID,
+	justificationBytes []byte,
+) *common.AppError {
+	justification, err := message.ParseRegisterSubnetValidator(justificationBytes)
+	if err != nil {
+		return &common.AppError{
+			Code:    ErrFailedToParseRegisterSubnetValidator,
+			Message: "failed to parse RegisterSubnetValidator justification: " + err.Error(),
+		}
+	}
+
+	justificationID := justification.ValidationID()
+	if validationID != justificationID {
+		return &common.AppError{
+			Code:    ErrMismatchedValidationID,
+			Message: fmt.Sprintf("validationID %q != justificationID %q", validationID, justificationID),
+		}
+	}
+
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	// Verify that the validator does not currently exist
+	_, err = s.state.GetSubnetOnlyValidator(validationID)
+	if err == nil {
+		return &common.AppError{
+			Code:    ErrValidationExists,
+			Message: fmt.Sprintf("validation %q exists", validationID),
+		}
+	}
+	if err != database.ErrNotFound {
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to lookup subnet only validator: " + err.Error(),
+		}
+	}
+
+	currentTimeUnix := uint64(s.state.GetTimestamp().Unix())
+	if justification.Expiry <= currentTimeUnix {
+		return nil // The expiry time has passed
+	}
+
+	// If the validation ID was successfully registered and then removed, it can
+	// never be re-used again even if its expiry has not yet passed.
+	hasExpiry, err := s.state.HasExpiry(state.ExpiryEntry{
+		Timestamp:    justification.Expiry,
+		ValidationID: validationID,
+	})
+	if err != nil {
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to lookup expiry: " + err.Error(),
+		}
+	}
+	if !hasExpiry {
+		return &common.AppError{
+			Code:    ErrValidationCouldBeRegistered,
+			Message: fmt.Sprintf("validation %q can be registered until %d", validationID, justification.Expiry),
+		}
+	}
+
+	return nil // The validator has been removed
+}
+
+func (s signatureRequestVerifier) verifySubnetValidatorWeight(
+	msg *message.SubnetValidatorWeight,
+) *common.AppError {
+	if msg.Nonce == math.MaxUint64 {
+		return &common.AppError{
+			Code:    ErrImpossibleNonce,
+			Message: "impossible nonce",
+		}
+	}
+
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	sov, err := s.state.GetSubnetOnlyValidator(msg.ValidationID)
+	switch {
+	case err == database.ErrNotFound:
+		return &common.AppError{
+			Code:    ErrValidationDoesNotExist,
+			Message: fmt.Sprintf("validation %q does not exist", msg.ValidationID),
+		}
+	case err != nil:
+		return &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: "failed to get subnet only validator: " + err.Error(),
+		}
+	case msg.Nonce+1 != sov.MinNonce:
+		return &common.AppError{
+			Code:    ErrWrongNonce,
+			Message: fmt.Sprintf("provided nonce %d != expected nonce (%d - 1)", msg.Nonce, sov.MinNonce),
+		}
+	case msg.Weight != sov.Weight:
```

Comment:
nit: should we check msg.Weight != 0 before? (to shortcircuit fetching from the state)

Reply:
I had considered doing this... The only reason we check the nonce is to prevent overflow... We could check this, but I felt like it didn't really prevent any DoSes... And it just adds more code.